### ISSUE TITLE
Modelo de predicción de riesgo de no-devolución (Clasificador)

### ISSUE THREAD

**Issue by Sebastian906** [Sebastian906/Zibooka-FullStack#181]
| Campo | Valor |
|-------|-------|
| **Tipo** | `feature` |
| **Prioridad** | 🟡 High |
| **Estimación** | 2 días |
| **Depende de** | #179 |
| **Bloquea** | Ninguno |

## Descripción

Al momento de crear un préstamo de libro (Loan), no hay forma de saber si el usuario tiene probabilidad de devolverlo tarde. Implementar un clasificador binario que estime la probabilidad de que un préstamo específico termine en estado `overdue`, basado en el historial del usuario, el historial del libro y el contexto temporal.

## Objetivo

Integrar la predicción en `LoanService.createLoan()` para:
1. Almacenar el `riskScore` en el documento Loan
2. Notificar al administrador si el riesgo es alto (> 0.7)
3. Opcionalmente, requerir confirmación adicional para préstamos de alto riesgo

## Features del Modelo

| Feature | Tipo | Descripción | Fuente |
|---------|------|-------------|--------|
| `user_previous_loans_count` | int | Total de préstamos del usuario | Aggregation Loan por userId |
| `user_overdue_count` | int | Veces que devolvió tarde (status=overdue) | Aggregation Loan por userId |
| `user_overdue_rate` | float | overdue_count / previous_loans_count | Derivado |
| `user_avg_late_days` | float | Días promedio de retraso (0 si nunca overdue) | Aggregation Loan por userId |
| `user_days_since_last_loan` | int | Días desde su último préstamo | Derivado de fechas |
| `user_total_loans_completed` | int | Préstamos completados exitosamente | Aggregation |
| `book_overdue_rate` | float | Tasa histórica de overdue del libro | Aggregation Loan por productId |
| `book_total_loans` | int | Préstamos totales del libro | Aggregation |
| `book_avg_loan_duration` | float | Duración promedio del préstamo del libro | Aggregation |
| `day_of_week_loan` | int (0-6) | Día de la semana del préstamo | Derivado de loanDate |
| `category_encoded` | int | Categoría del libro (label encoded) | Product.category |
| `is_weekend` | bin (0/1) | 1 si el préstamo se hace en sábado o domingo | Derivado |
| `loan_hour` | int (0-23) | Hora del día del préstamo | loanDate.getHours() |

## Target

Variable binaria: **1** si el préstamo terminó con `status === 'overdue'`, **0** si `status === 'returned'` con `returnDate <= dueDate`.

### 1. Clase OverduePredictor

**Archivo:** `zibooka-ml/app/models/overdue_predictor.py`

```python
import pandas as pd
import numpy as np
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import classification_report, roc_auc_score, confusion_matrix
import joblib
from app.models.base import BasePredictor
from app.database import mongodb

class OverduePredictor(BasePredictor):
    def __init__(self):
        self.model = XGBClassifier(
            n_estimators=150,
            max_depth=5,
            learning_rate=0.1,
            random_state=42,
            scale_pos_weight=2,  # Balancear clases (overdue es minoría)
            eval_metric='logloss'
        )
        self.feature_columns = [
            'user_previous_loans_count', 'user_overdue_count', 'user_overdue_rate',
            'user_avg_late_days', 'user_days_since_last_loan', 'user_total_loans_completed',
            'book_overdue_rate', 'book_total_loans', 'book_avg_loan_duration',
            'day_of_week_loan', 'category_encoded', 'is_weekend', 'loan_hour'
        ]
        self.label_encoder = {}
    
    async def _build_training_data(self) -> pd.DataFrame:
        """Construye dataset con features de usuario + libro + contexto."""
        pipeline = [
            {
                '$lookup': {
                    'from': 'products',
                    'localField': 'productId',
                    'foreignField': '_id',
                    'as': 'product'
                }
            },
            { '$unwind': '$product' },
            {
                '$lookup': {
                    'from': 'users',
                    'localField': 'userId',
                    'foreignField': '_id',
                    'as': 'user'
                }
            },
            { '$unwind': '$user' },
            # Solo préstamos con resolución conocida
            { '$match': { 'status': { '$in': ['returned', 'overdue'] } } },
            {
                '$project': {
                    'userId': 1,
                    'productId': 1,
                    'status': 1,
                    'loanDate': 1,
                    'dueDate': 1,
                    'returnDate': 1,
                    'category': '$product.category',
                    'userName': '$user.name'
                }
            }
        ]
        
        loans_cursor = mongodb.get_collection('loans').aggregate(pipeline)
        loans = await loans_cursor.to_list(length=None)
        return pd.DataFrame(loans)
    
    def _engineer_features(self, df: pd.DataFrame) -> pd.DataFrame:
        """Construye feature vector para cada préstamo."""
        features = df.copy()
        
        # Label encoding para categorías
        categories = features['category'].unique()
        self.label_encoder = {cat: i for i, cat in enumerate(categories)}
        features['category_encoded'] = features['category'].map(self.label_encoder)
        
        # Features temporales
        features['loanDate'] = pd.to_datetime(features['loanDate'])
        features['day_of_week_loan'] = features['loanDate'].dt.dayofweek
        features['is_weekend'] = features['day_of_week_loan'].isin([5, 6]).astype(int)
        features['loan_hour'] = features['loanDate'].dt.hour
        
        # Target: overdue = 1, returned = 0
        features['target'] = (features['status'] == 'overdue').astype(int)
        
        # Las features agregadas por usuario y libro se calculan
        # con ventanas temporales (excluyendo el préstamo actual)
        # ... (lógica de agregación por userId y productId)
        
        return features
    
    def train(self, training_data: pd.DataFrame) -> dict:
        features = self._engineer_features(training_data)
        
        X = features[self.feature_columns]
        y = features['target']
        
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=0.2, random_state=42, stratify=y
        )
        
        # GridSearch para hiperparámetros
        param_grid = {
            'max_depth': [3, 5, 7],
            'n_estimators': [100, 150, 200],
            'learning_rate': [0.05, 0.1, 0.2]
        }
        
        grid_search = GridSearchCV(
            XGBClassifier(random_state=42, eval_metric='logloss'),
            param_grid,
            cv=5,
            scoring='roc_auc',
            n_jobs=-1
        )
        
        grid_search.fit(X_train, y_train)
        self.model = grid_search.best_estimator_
        
        # Evaluación
        y_pred = self.model.predict(X_test)
        y_prob = self.model.predict_proba(X_test)[:, 1]
        
        tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()
        
        metrics = {
            'best_params': grid_search.best_params_,
            'accuracy': float((tp + tn) / (tp + tn + fp + fn)),
            'precision': float(tp / (tp + fp)) if (tp + fp) > 0 else 0,
            'recall': float(tp / (tp + fn)) if (tp + fn) > 0 else 0,
            'f1_score': float(2 * tp / (2 * tp + fp + fn)) if (2 * tp + fp + fn) > 0 else 0,
            'auc_roc': float(roc_auc_score(y_test, y_prob)),
            'true_negatives': int(tn),
            'false_positives': int(fp),
            'false_negatives': int(fn),
            'true_positives': int(tp)
        }
        
        return metrics
    
    def predict(self, features: np.ndarray) -> np.ndarray:
        """Retorna probabilidad de overdue (0-1)."""
        return self.model.predict_proba(features)[:, 1]
    
    def predict_single(self, user_id: str, book_id: str) -> float:
        """Predice riesgo para un par usuario-libro específico."""
        # Construir feature vector en tiempo real
        # ...
        pass
    
    def save(self, path: str) -> None:
        joblib.dump({
            'model': self.model,
            'feature_columns': self.feature_columns,
            'label_encoder': self.label_encoder
        }, path)
    
    @classmethod
    def load(cls, path: str) -> 'OverduePredictor':
        data = joblib.load(path)
        predictor = cls()
        predictor.model = data['model']
        predictor.feature_columns = data['feature_columns']
        predictor.label_encoder = data['label_encoder']
        return predictor
    
    def get_feature_importance(self) -> dict:
        importance = self.model.feature_importances_
        return dict(zip(self.feature_columns, importance.tolist()))
```

### 2. Integración en NestJS — LoanService

**Archivo:** `server/src/loans/loan.service.ts`

```typescript
@Injectable()
export class LoanService {
  constructor(
    // ... dependencias existentes
    private predictionClient: PredictionClient,
    @InjectModel(Loan.name) private loanModel: Model<LoanDocument>,
  ) {}
  
  async createLoan(userId: string, productId: string): Promise<Loan> {
    // ... validaciones existentes
    
    // Llamar al microservicio ML
    const riskScore = await this.predictionClient.predictOverdueRisk(userId, productId);
    
    // Crear préstamo con riskScore
    const loan = new this.loanModel({
      userId,
      productId,
      loanDate: new Date(),
      status: 'active',
      riskScore: riskScore,
    });
    
    await loan.save();
    
    // Si el riesgo es alto, notificar al admin
    if (riskScore > 0.7) {
      await this.notifyHighRiskLoan(loan, riskScore);
    }
    
    return loan;
  }
  
  private async notifyHighRiskLoan(loan: LoanDocument, riskScore: number): Promise<void> {
    // Enviar notificación al admin (email, panel, etc.)
    await this.emailService.sendAlert({
      to: process.env.ADMIN_EMAIL,
      subject: 'Préstamo de Alto Riesgo Detectado',
      template: 'high-risk-loan',
      context: {
        loanId: loan._id,
        userId: loan.userId,
        productId: loan.productId,
        riskScore: riskScore,
        date: loan.loanDate
      }
    });
    
    // También registrar en una colección de alertas
    await this.highRiskLoanModel.create({
      loanId: loan._id,
      riskScore,
      notifiedAt: new Date()
    });
  }
}
```

### 3. Schema — Agregar riskScore a Loan

**Archivo:** `server/src/loans/schemas/loan.schema.ts`

```typescript
@Schema({ timestamps: true })
export class Loan {
  // ... campos existentes
  
  @Prop({ type: Number, min: 0, max: 1, default: 0.5 })
  riskScore: number;  // Probabilidad de overdue (0-1)
}
```

### 4. Nuevo Schema — HighRiskLoan

**Archivo:** `server/src/loans/schemas/high-risk-loan.schema.ts`

```typescript
@Schema({ timestamps: true })
export class HighRiskLoan {
  @Prop({ type: mongoose.Schema.Types.ObjectId, ref: 'Loan', required: true })
  loanId: string;
  
  @Prop({ required: true })
  riskScore: number;
  
  @Prop({ default: false })
  reviewed: boolean;  // Si el admin ya revisó la alerta
  
  @Prop()
  reviewedAt: Date;
  
  @Prop()
  adminNotes: string;
}
```

## Criterios de Aceptación

- [ ] El clasificador tiene AUC-ROC > 0.75 en datos de prueba
- [ ] `POST /train/overdue` entrena y guarda el modelo con GridSearch
- [ ] `POST /predict/overdue` retorna `{ risk_score, top_features }`
- [ ] `LoanService.createLoan()` integra la llamada al microservicio
- [ ] Si el microservicio falla, la creación del préstamo continúa (fallback a 0.5)
- [ ] El `riskScore` se persiste en el documento Loan
- [ ] Préstamos con riskScore > 0.7 generan una alerta revisable por el admin
- [ ] Hay un panel admin para ver "Préstamos de Alto Riesgo"
- [ ] El modelo maneja desbalanceo de clases (scale_pos_weight)

## Límites y Fallback

```typescript
// En PredictionClient
async predictOverdueRisk(userId: string, bookId: string): Promise<number> {
  try {
    const response = await fetch(`${this.baseUrl}/predict/overdue`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ user_id: userId, book_id: bookId }),
      signal: AbortSignal.timeout(3000)  // Timeout 3s
    });
    const data = await response.json();
    return data.risk_score;
  } catch (error) {
    // Fallback: riesgo neutral
    console.warn('[PredictionClient] Overdue predictor unavailable, using default:', error.message);
    return 0.5;
  }
}
```
